### ISSUE TITLE
Add String.Format overloads to avoid unnecessary allocations

### ISSUE THREAD

**Issue by justinvp** [dotnet/corefx#1514]
It's pretty common to pass value types to `String.Format`, unfortunately this results in unnecessary boxing allocations.

## Rationale and Usage

Consider the following:

```c#
long id = ...;
int index = ...;
int length = ...;

// Traditional call to String.Format
string foo = string.Format("{0}: Index = {1}, Length = {2}", id, index, length);

// Use of the string interpolation language feature
string bar = $"{id}: Index = {index}, Length = {length}";
```

Both the traditional call to `String.Format` and the use of the string interpolation language feature (which is just syntactic sugar for `String.Format`) requires 3 boxing allocations.

## Proposed API

```c#
namespace System
{
    public sealed class String : ...
    {
        // Proposed methods
        public static string Format<T>(string format, T arg);
        public static string Format<T1, T2>(string format, T1 arg1, T2 arg2);
        public static string Format<T1, T2, T3>(string format, T1 arg1, T2 arg2, T3 arg3);
        public static string Format<T>(IFormatProvider provider, string format, T arg);
        public static string Format<T1, T2>(IFormatProvider provider, string format, T1 arg1, T2 arg2);
        public static string Format<T1, T2, T3>(IFormatProvider provider, string format, T1 arg1, T2 arg2, T3 arg3);
    
        // Existing methods
        public static string Format(string format, object arg0);
        public static string Format(string format, object arg0, object arg1);
        public static string Format(string format, object arg0, object arg1, object arg2);
        public static string Format(string format, params object[] args);
        public static string Format(IFormatProvider provider, string format, object arg0);
        public static string Format(IFormatProvider provider, string format, object arg0, object arg1);
        public static string Format(IFormatProvider provider, string format, object arg0, object arg1, object arg2);
        public static string Format(IFormatProvider provider, string format, params object[] args);
        
        ...
    }
}

namespace System.Text
{
    public sealed class StringBuilder
    {
        // Proposed methods
        public StringBuilder AppendFormat<T>(string format, T arg);
        public StringBuilder AppendFormat<T1, T2>(string format, T1 arg1, T2 arg2);
        public StringBuilder AppendFormat<T1, T2, T3>(string format, T1 arg1, T2 arg2, T3 arg3);
        public StringBuilder AppendFormat<T>(IFormatProvider provider, string format, T arg);
        public StringBuilder AppendFormat<T1, T2>(IFormatProvider provider, string format, T1 arg1, T2 arg2);
        public StringBuilder AppendFormat<T1, T2, T3>(IFormatProvider provider, string format, T1 arg1, T2 arg2, T3 arg3);
    
        // Existing methods
        public StringBuilder AppendFormat(string format, object arg0);
        public StringBuilder AppendFormat(string format, object arg0, object arg1);
        public StringBuilder AppendFormat(string format, object arg0, object arg1, object arg2);
        public StringBuilder AppendFormat(string format, params object[] args);
        public StringBuilder AppendFormat(IFormatProvider provider, string format, object arg0);
        public StringBuilder AppendFormat(IFormatProvider provider, string format, object arg0, object arg1);
        public StringBuilder AppendFormat(IFormatProvider provider, string format, object arg0, object arg1, object arg2);
        public StringBuilder AppendFormat(IFormatProvider provider, string format, params object[] args);
        
        ...
    }
}

namespace System.Runtime.CompilerServices
{
    public static class FormattableStringFactory
    {
        // Proposed methods
        public static FormattableString Create<T>(string format, T arg);
        public static FormattableString Create<T1, T2>(string format, T1 arg1, T2 arg2);
        public static FormattableString Create<T1, T2, T3>(string format, T1 arg1, T2 arg2, T3 arg3);
        
        // Existing method
        public static FormattableString Create(string format, params object[] arguments);
    }
}
```

## Details

TBD

## Questions

 - Should the argument names be `arg1`, `arg2`, `arg3` (consistent with `Action` and `Func`) or spelled out as `argument1`, `argument2`, `argument3`?

I'd be happy to contribute an implementation and tests.

**Comment by ellismg** [comment]
Do the additional FormattableStringFactory members require Roslyn changes to be consumed with the nice syntax?

**Comment by svick** [comment]
I think the parameters in the new overloads should have exactly the same names as in the old overloads, so that if someone called `string.Format()` with named parameters, their code also starts using the new overloads just by recompiling.

**Comment by justinvp** [comment]
> Do the additional FormattableStringFactory members require Roslyn changes to be consumed with the nice syntax?

I need to double-check, but I'm pretty sure Roslyn is already designed to make use of such overloads to `String.Format` and `FormattableStringFactory.Create`, if they exist. So, no Roslyn changes would be required.

**Comment by justinvp** [comment]
> I think the parameters in the new overloads should have exactly the same names as in the old overloads

The 0-based parameter names seem a little strange, but now that you mention keeping the naming the same as the old overloads, the format items in the format string are 0-based, so it makes sense that the parameter names match that. In that case, I'm guessing we'd want the naming of the generic type parameters to match the naming of the arguments:

So instead of this:

```c#
public static string Format<T>(string format, T arg);
public static string Format<T1, T2>(string format, T1 arg1, T2 arg2);
public static string Format<T1, T2, T3>(string format, T1 arg1, T2 arg2, T3 arg3);
```

This:

```c#
public static string Format<T0>(string format, T0 arg0);
public static string Format<T0, T1>(string format, T0 arg0, T1 arg1);
public static string Format<T0, T1, T2>(string format, T0 arg1, T1 arg1, T2 arg2);
```

Or maybe even `TArgument0`, `TArgument1`, and `TArgument2` vs. `T0`, `T1`,  and `T2`.

**Comment by weshaggard** [comment]
We did talk about the generic overloads on FormattableStringFactory as part of the review but we didn't believe we had enough runway to get the compiler to actually use those in this product release. (cc @gafter). 

As for the string.Format overloads we did end up doing some work this release to make the arg0/arg1/arg2 overloads actually avoid the params array allocation internally (yeah until recently we didn't actually even optimize that even though we had the overloads). While we did do that one thing we realized is that string.Format is extremely bad about allocations internally and we don't believe that even eliminating the params array allocation ended up improving performance as much as we would have liked it to. I'm primarily saying that because I have to wonder if a few boxing allocations is really going to have much of an impact when what is really needed is a much larger redoing of the underlying formatting code (which I'm not sure is even possible in a compatible way). Do you have any data about whether or not eliminating these boxing allocations will end up helping much? By the way this doesn't mean these aren't worth adding I'm just trying to collect some more data.

**Comment by gafter** [comment]
The C# and VB compilers will use appropriate overloads on `FormattableStringFactory` if those exist. What we didn't have enough runway for was writing those overloads.

**Comment by pharring** [comment]
My experience is that the inner workings of string.Format (actually StringBuilder.AppendFormat) is quite efficient. The allocations are what you'd expect. Most of the time we get a hit in StringBuilderCache, so there's no char[] allocation.
Allocations in this area usually show up in serialization and logging routines. Serialization usually arrives via TextWriter.Write (which calls string.Format).
Removing the params array in 4.6 was nice.
Roslyn analyzers can help locate and eliminate boxing.

It would be interesting to see if one could implement this, avoiding boxing but at the same time avoiding stamping out 3 new generic versions of StringBuilder.AppendFormat.

**Comment by mattwarren** [comment]
> Roslyn analyzers can help locate and eliminate boxing.

There's already [an analyser](https://github.com/mjsabby/RoslynClrHeapAllocationAnalyzer/) that does this. The tricky part is keeping it up to date with the changes, for instance it currently [doesn't handle](https://github.com/mjsabby/RoslynClrHeapAllocationAnalyzer/issues/8) the recent changes that got rid of [unnecessary boxing in String.Concat](https://github.com/dotnet/roslyn/pull/415). Probably the same applies for the change that removed params array allocations, that has been mentioned elsewhere in this thread.

The robust way to do this would be for it to look at the IL produced, but that's going to be expensive for an Analyser (although I did give it a go and Roslyn combined with Cecil makes it nice and easy).

**Comment by justinvp** [comment]
> Do you have any data about whether or not eliminating these boxing allocations will end up helping much?

@weshaggard, I have some data in the conclusion below, but first some background...

Consider the following example:

```c#
DateTime now = DateTime.UtcNow;
Guid id = Guid.NewGuid();
decimal value = 3.50m;

string result = string.Format("{0:s} - {1:B}: The value is: {2:C2}", now, id, value);
```

Let's track of the allocations with the current implementation:

 1. The `now` parameter is boxed.
 1. The `id` parameter is boxed.
 1. The `value` parameter is boxed.
 1. A new `StringBuilder` is allocated to build-up the `"s"` (in `"{0:s}"`) format string.
 1. The actual `"s"` `string` is allocated.
 1. The resulting string for `now.ToString("s", null)` is allocated.
 1. A new `StringBuilder` is allocated to build-up the `"B"` (in `"{1:B}"`) format string.
 1. The actual `"B"` `string` is allocated.
 1. The resulting string for `id.ToString("B", null)` is allocated.
 1. A new `StringBuilder` is allocated to build-up the `"C2"` (in `"{2:C2}"`) format string.
 1. The actual `"C2"` `string` is allocated.
 1. The resulting string for `value.ToString("C2", null)` is allocated.
 1. The overall resulting `string` is allocated.

13 allocations!

Note that there's also a `StringBuilder` instance used for the overall resulting string that I left out of the allocation count as this allocation is mitigated by the use of the internal `StringBuilderCache`.

These allocations can be reduced significantly!

 - A) Generic overloads would reduce this by 3, down to 10 allocations.

 - B) There is some low-hanging fruit inside the `String.Format` (`StringBuilder.AppendFormat`) implementation that would avoid the `StringBuilder` allocations for the argument format strings, using `string.Substring` for the common case and falling back to a single `StringBuilder` instance only when necessary (more on this below), which would reduce the allocations by another 3, down to 7 allocations.

 - C) Using an alternative design to `IFormattable` (even if just used internally for the built-in BCL types) could reduce this by another 6 allocations, down to just 1 allocation!

Down to just 1 allocation for the resulting string!

### A) Generic overloads

Avoids the boxing allocations. Pretty straightforward.

### B) Low-hanging fruit (avoiding/mitigating the internal `StringBuilder` allocations)

Right now with the current implementation, a new `StringBuilder` is allocated for each argument that has a format string associated with it (e.g. the `"C2"` in `"{0:C2}"`). This can be trivially improved by simply reusing the `StringBuilder` instance for multiple args. It involves just moving the `StringBuilder` variable to the top of the loop, still lazily allocating it, clearing it when finished, and reusing it for the next argument that needs it. However, even better, in the common case when there aren't any escaped braces in the argument's format string (e.g. `"{{"` and `"}}"`) then a `StringBuilder` isn't needed at all; the more light-weight `String.Substring` can be used.

### C) `IFormattable` alternative

Optimizing `String.Format` (`StringBuilder.AppendFormat`) is ultimately limited by the design of `IFormattable`, and to a lesser extent the less commonly used `ICustomFormatter`.

```c#
public interface IFormattable
{
    string ToString(string format, IFormatProvider formatProvider);
}
```

This design is perfectly reasonable, but there are a couple issues that prevent optimizations inside `String.Format` (`StringBuilder.AppendFormat`):

 1. The `format` parameter is typed as `string`, which means an instance of `string` must be allocated for `format`, as opposed to accepting a `StringSpan`.

 2. The result is a `string`, which means an instance of `string` must be allocated for the result, as opposed to appending the resulting characters to a destination `StringBuilder`.

An alternative design might look something like the following:

```c#
public interface IFormattableTo // (no pun intended, heh)
{
    void FormatTo(StringSpan format, IFormatProvider formatProvider, StringBuilder destination);
}
```

It may not be worth exposing such an interface publicly (at least for now) as it would basically be a replacement for `IFormattable` and it requires the `StringSpan` or `Span<char>` type, which hasn't landed yet. But it could be included as internal for now and implemented on all the built-in BCL types that currently implement `IFormattable` for improved `String.Format` perf for these very commonly used types.

## Conclusion

Microbenchmarking the above example with changes A and B implemented (10,000,000 iterations), I'm seeing a ~3% perf improvement over the current implementation (in addition to the reduction in heap allocations). And by changing how the implementation loops through the characters in the format string (changing it to use a `char*` pointer) I see an additional ~1% perf improvement, for a ~4% perf improvement.

I haven't had a chance to dive into C in detail yet, but I think it'd be worth at least investigating further.

**Comment by MikePopoloski** [comment]
Shameless plug of my own work: https://github.com/MikePopoloski/StringFormatter
Might be useful to mine for ideas.

I too needed string formatting without allocations. It'd be nice to take advantage of some of that dedicated code built into the CLR to make it faster, but it does alright on its own as pure C#.

**Comment by justinvp** [comment]
@terrajobst, Immo, this is an API suggestion. Could this be added to the agenda for an API review? Even if there's some opposition to the `String` and `StringBuilder` overloads, I think the `FormattableStringFactory` overloads are a no-brainer (would have been added for 4.6, but ran out of time).

**Comment by chintsan** [comment]
Hi Justin, 
thanks for the explanation ans benchmarks.

how does one include the above overloads. i mean defining a new class "System" throws error - no body implementation for the overload methods or they arent defined as abstract, partial, extern etc.

**Comment by NickCraver** [comment]
I know it's been a while, but we'd love to see this over at Stack Overflow - @terrajobst can you please ping here when the API review is up next so we can listen in?

**Comment by terrajobst** [comment]
@justinvp 

> Could this be added to the agenda for an API review?

Generally yes, I believe with have everything we need. However, we're currently focusing on getting RTM ready, which means we're focusing our reviews on exposing APIs that already exist in .NET Framework to simplify porting as well as adding new APIs that are required to unblock other components and tools. That's why it's currently in the `Future` milestone. I hope that we'll be able to move some of these to a post-RTM milestone.

**Comment by AdamSpeight2008** [comment]
We could get the maximum `.ToString()` of most of core types. Eg `.MaxValue.ToString.Length + 1` 
```c#
ApproxMaxCharLength<T>( obj : T ) : Int
{ 
  return match( typeof( T ) ) with
  {
    | string -> ((string)obj).Length; 
    |   bool ->  5; //  false
    |   byte ->  3; //  256
    |  sbyte ->  4; // -128
    |   char ->  1; // 
    | uint16 ->  6; //  65535
    |  int16 ->  6; // -32768
    | uint32 -> 11; // -2147483648 
    |  int32 -> 10; //  4294967295
    | uint64 -> 20; //  18446744073709551615
    |  int64 -> 20; // –9223372036854775808
    |    ... ->  0;
    // not done the floating point ones.
  }
}  
```
An initial first approximation of the string length, could be calculated thus.
```c#

var intialSize = args.Sum( ApproxMaxCharLength ) + TextOfFormatString.Length; 
```
I **think** this would be an over allocation. **Note** this is exclude any variation caused by the culture and additional alignment and format specifiers on the arg hole eg `{ x , -16 : X4 }`

At which point *(don't done any deep analysis)* would the extra processing to get a good estimate of the size of string to allocation, outweighs the processing cost of just blindly done the building of the string.

**Comment by gafter** [comment]
@AdamSpeight2008 This issue is mainly about the allocations due to boxing.

**Comment by terrajobst** [comment]
The APIs look fine, however, the only way we could benefit from this work is if our implementation avoids the boxing all the way down, which would be non-trivial amount of work. Also, it's questionable without a benchmark how big the savings are, because formatting will probably cause a bunch of allocations anyway.

It might be better to leave these APIs as-is and instead use the new formatting APIs that @KrzysztofCwalina is working on.

@KrzysztofCwalina, what are your thoughts on this?

**Comment by NickCraver** [comment]
Remember that the benefits here are extreme because *just by recompiling*, you get them. That huge, free adoption should be factored into the effort costing. This helps all code, rather than just new code. I don't see `string.Format()` or `$""` going anywhere for a long time. 

Perhaps I'm missing something though - doesn't `$""` continue to map to `string.Format` in future plans as well, or is that being re-pointed to some new API? That'd be a pretty extreme breaking change, but worth asking...

**Comment by gafter** [comment]
`$"..."` always maps to an invocation of `string.Format` (if the target type is `string`).

**Comment by KrzysztofCwalina** [comment]
I like the improvement in-place for the reson @NickCraver spells out. I think we should add these APIs.
The only concern I would have is that it might cause Generics instantiation explosion that runtime people are really concerned about lately. @jkotas?

**Comment by jkotas** [comment]
This should be data-driven decision. It is unclear to me whether saving extra allocations is worth it to justify the extra code in this case. String.Format will always allocate and it will always be relatively slow API. It is unclear to me whether changing it to allocate a bit less garbage at the cost of pile of extra code is going to measurable improve anything important.

**Comment by benaadams** [comment]
The trouble is string interpolation is too good a feature not to use, even for those of use that try to avoid allocations at all costs...

**Comment by jkotas** [comment]
Agree. In the above @justinvp example, there 13 allocations. 9 of them are internal. It should be possible to eliminate all internal ones without changing the shape of the API, so that should be done first.

**Comment by karelz** [comment]
Next step: Improve current memory allocations of existing APIs as much as possible -- the result will tell us if we still need new APIs.

**Comment by NickCraver** [comment]
How would we *ever* eliminate the boxing here without the API additions? I agree with internal eliminations being the first step, but if we're after absolute performance, I don't see a way around the second step.

**Comment by jkotas** [comment]
> we're after absolute performance, I don't see a way around the second step.

The no-allocation formatters being developed in corefxlab are going after absolute performance. 

`String.Format` will always allocate the string. The string is likely short-lived garbage - it is unlikely that it is stored for long. I do not think that the difference between allocating a string vs. allocating a string and a box for each argument is going to be that significant.

**Comment by JonHanna** [comment]
(Aside regarding the original proposed API rather than the current discussion, would `public static string Format<T>(string format, params T[] args);` be useful or overkill? It's not rare to have to output a bunch of integers or a bunch of dates or otherwise format several items of the same type).

**Comment by NickCraver** [comment]
@jkotas I agree on `String.Format()`, but our much more common use case (localization in views, etc.) is on the `StringBuilder.AppendFormat()` side of the fence, which has more to win here overall.  An example of this is in ASP.NET views, especially when we're doing localization. 

Either way, we need to measure. The trouble is measuring this accurately means really doing it for comparison...to see if it's worth doing. Vicious circle we have here :)

**Comment by jkotas** [comment]
> The trouble is measuring this accurately means really doing it for comparison

Right, and doing it is non-trivial amount of work because of the current formatting implementation has heavy dependency on boxing.

**Comment by george-polevoy** [comment]
The new `Format(Span<char> span, out int length)` family of functions is what makes the fastest implementation possible, using zero alloc techniques.

Rewrite `$"prefix{x}postfix"`

to

```
var resultBuffer = stackalloc new char[<just enough for the value types in question> plus the string constant];
var buffer = resultBuffer;
int length;
"prefix".CopyTo(buffer);
buffer = buffer.Slice("prefix".Length);
arg1.Format(buffer, out length);
buffer = buffer.Slice(length);
"postfix".CopyTo(buffer);
resultBuffer.Overlaps(buffer, out var elementOffset);
return new string(resultBuffer.Slice(0, elementOffset));
```

I understand that this would totally break compatibility, so this could be an extension to a future C# version, like:
```
$ <span expression> "{x}"
```
or 
```
$ stackalloc char[] "{x}"
$ stackalloc utf8 "{x}"
```

Another possibility to avoid boxing would be to call ToString() on each of the arguments in-place, and also provide an overload which takes a span of strings instead of array of objects.
